File: computing/interview/design-patterns/behavioral/iterator/java/IteratorDemo.java
package interview.designpatterns.behavioral.iterator;

import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.concurrent.CopyOnWriteArrayList;

final class Playlist implements Iterable<String> {
    private final List<String> tracks = new CopyOnWriteArrayList<>();

    void addTrack(String name) {
        tracks.add(name);
    }

    @Override
    public Iterator<String> iterator() {
        return new PlaylistIterator(tracks);
    }

    private static final class PlaylistIterator implements Iterator<String> {
        private final List<String> tracks;
        private int index;

        PlaylistIterator(List<String> tracks) {
            this.tracks = tracks;
        }

        @Override
        public boolean hasNext() {
            return index < tracks.size();
        }

        @Override
        public String next() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }
            return tracks.get(index++);
        }
    }
}

public final class IteratorDemo {
    private IteratorDemo() {}

    public static void main(String[] args) {
        Playlist playlist = new Playlist();
        playlist.addTrack("Intro");
        playlist.addTrack("Theme");
        playlist.addTrack("Outro");

        for (String track : playlist) {
            System.out.printf("Track: %s%n", track);
        }
    }
}

File: computing/interview/design-patterns/behavioral/iterator/python/iterator_demo.py
from __future__ import annotations
from collections.abc import Iterable, Iterator
from typing import List


class Playlist(Iterable[str]):
    def __init__(self) -> None:
        self._tracks: List[str] = []

    def add_track(self, name: str) -> None:
        self._tracks.append(name)

    def __iter__(self) -> Iterator[str]:
        return _PlaylistIterator(self._tracks)


class _PlaylistIterator(Iterator[str]):
    def __init__(self, tracks: List[str]) -> None:
        self._tracks = tracks
        self._index = 0

    def __next__(self) -> str:
        if self._index >= len(self._tracks):
            raise StopIteration
        value = self._tracks[self._index]
        self._index += 1
        return value


def main() -> None:
    playlist = Playlist()
    playlist.add_track("Intro")
    playlist.add_track("Theme")
    playlist.add_track("Outro")

    for track in playlist:
        print(f"Track: {track}")


if __name__ == "__main__":
    main()
